bfvmm/src/hve/arch/intel_x64/ept/memory_map.cpp

```diff
@@ -0,0 +1,340 @@
+            expects(pte::page_address::is_aligned(hpa));
+            return map_pte_to_page(gpa, hpa);
+
+        default:
+            throw std::logic_error("map: invalid ept page size specified");
+    }
+}
+
+void
+memory_map::unmap(gpa_t gpa) const
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    epte_t &leaf = gpa_to_epte(gpa);
+    epte::clear(leaf);
+}
+
+epte_t &
+memory_map::gpa_to_epte(gpa_t gpa) const
+{
+    auto err = "gpa_to_epte: failed to resolve gpa->epte, gpa is not mapped";
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    epte_t &pml4e = gpa_to_pml4e(gpa);
+    if (!epte::is_present(pml4e)) { throw std::runtime_error(err); }
+
+    epte_t &pdpte = gpa_to_pdpte(gpa, pml4e);
+    if (!epte::is_present(pdpte)) { throw std::runtime_error(err); }
+    if (epte::is_leaf_entry(pdpte)) { return pdpte; }
+
+    epte_t &pde = gpa_to_pde(gpa, pdpte);
+    if (!epte::is_present(pde)) { throw std::runtime_error(err); }
+    if (epte::is_leaf_entry(pde)) { return pde; }
+
+    epte_t &pte = gpa_to_pte(gpa, pde);
+    if (!epte::is_present(pte)) { throw std::runtime_error(err); }
+    if (epte::is_leaf_entry(pte)) { return pte; }
+
+    throw std::runtime_error("gpa_to_epte: extended page tables corrupted");
+}
+
+hpa_t
+memory_map::gpa_to_hpa(gpa_t gpa) const
+{
+    auto err = "gpa_to_hpa: failed to resolve gpa->hpa, gpa is not mapped";
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    epte_t &pml4e = gpa_to_pml4e(gpa);
+    if (!epte::is_present(pml4e)) { throw std::runtime_error(err); }
+
+    epte_t &pdpte = gpa_to_pdpte(gpa, pml4e);
+    if (!epte::is_present(pdpte)) { throw std::runtime_error(err); }
+    if (epte::is_leaf_entry(pdpte)) {
+        return pdpte::page_address::get_effective_address(pdpte, gpa);
+    }
+
+    epte_t &pde = gpa_to_pde(gpa, pdpte);
+    if (!epte::is_present(pde)) { throw std::runtime_error(err); }
+    if (epte::is_leaf_entry(pde)) {
+        return pde::page_address::get_effective_address(pde, gpa);
+    }
+
+    epte_t &pte = gpa_to_pte(gpa, pde);
+    if (!epte::is_present(pte)) { throw std::runtime_error(err); }
+    if (epte::is_leaf_entry(pte)) {
+        return pte::page_address::get_effective_address(pte, gpa);
+    }
+
+    throw std::runtime_error("gpa_to_epte: extended page tables corrupted");
+}
+
+std::vector<memory_descriptor>
+memory_map::to_mdl() const
+{
+    std::vector<memory_descriptor> mdl{0};
+    auto &&type = MEMORY_TYPE_R | MEMORY_TYPE_W;
+    mdl.push_back({m_pml4_hpa, m_pml4_hva, type});
+
+    mdl = this->to_mdl(mdl, reinterpret_cast<epte_t *>(m_pml4_hva));
+    return mdl;
+}
+
+hpa_t
+memory_map::hpa() const
+{
+    return m_pml4_hpa;
+}
+
+hpa_t
+memory_map::allocate_page_table() const
+{
+    epte_t *pt_hva = new epte_t [page_table::num_entries] {0};
+    hpa_t pt_hpa = g_mm->virtptr_to_physint(pt_hva);
+    return pt_hpa;
+}
+
+void
+memory_map::allocate_page_table(epte_t &entry) const
+{
+    expects(entry == 0);
```

Comment:
bad idea. expects should only be used on public APIs, and you should set this to 0 yourself. Code like this would break on a simple error resulting in instability. Instead, be proactive, and make sure your code is in the state that you want and "fight back" on errors as you can. An error like this really should be a warning if you want to print anything at all (like bfwarning(1)). 

Fix all as needed. Remember..... try to survive as many issues as possible, and throw where you cannot do anything else.